File: adapters/python/run_chen.py
#!/usr/bin/env python3
"""chen-signatures adapter for signature benchmarks"""

import json
import sys
from typing import Any, Callable, Dict, Optional

import numpy as np
from common import BenchmarkAdapter, make_path


class ChenSignaturesAdapter(BenchmarkAdapter):
    """Adapter for chen-signatures library"""

    def __init__(self, config: Dict[str, Any]):
        super().__init__(config)
        # Import here to avoid import errors if not available
        import chen
        self.chen = chen

    def run_signature(self, path: np.ndarray, d: int, m: int) -> Optional[Callable]:
        """
        Prepare signature computation kernel.

        Returns a closure that performs only the kernel (no setup).
        """
        # Setup phase (untimed): ensure path is contiguous
        path = np.ascontiguousarray(path, dtype=np.float64)

        # Return kernel closure
        return lambda: self.chen.sig(path, m)

    def run_logsignature(self, path: np.ndarray, d: int, m: int) -> Optional[Callable]:
        """
        Prepare logsignature computation kernel.

        Returns a closure that performs only the kernel (no setup).
        """
        # Check if logsig methods are available
        if not (hasattr(self.chen, "logsig") and hasattr(self.chen, "prepare_logsig")):
            return None

        # Setup phase (untimed): prepare basis and ensure path is contiguous
        path = np.ascontiguousarray(path, dtype=np.float64)
        basis = self.chen.prepare_logsig(d, m)

        # Return kernel closure
        return lambda: self.chen.logsig(path, basis)

    def run_sigdiff(self, path: np.ndarray, d: int, m: int) -> Optional[Callable]:
        """
        Prepare signature differentiation kernel.

        Returns a closure that performs only the kernel (no setup).
        """
        # Import PyTorch dependencies
        try:
            from chen.torch import sig_torch
            import torch
        except ImportError:
            return None

        # Setup phase (untimed): prepare path as numpy array
        path_np = np.ascontiguousarray(make_path(d, self.N, self.path_kind), dtype=np.float64)

        # Return kernel closure that converts to torch, computes sig, and backprop
        def kernel():
            path_t = torch.tensor(path_np, dtype=torch.float64, requires_grad=True)
            sig = sig_torch(path_t, m)
            loss = sig.sum()
            loss.backward()

        return kernel

    def _run_benchmark(self) -> Optional[Dict[str, Any]]:
        """Execute the benchmark"""
        # Generate path
        path = make_path(self.d, self.N, self.path_kind)

        # Select operation
        if self.operation == "signature":
            kernel = self.run_signature(path, self.d, self.m)
            method = "sig"
            path_type = "ndarray"
        elif self.operation == "logsignature":
            kernel = self.run_logsignature(path, self.d, self.m)
            method = "logsig(prepared)"
            path_type = "ndarray"
        elif self.operation == "sigdiff":
            kernel = self.run_sigdiff(path, self.d, self.m)
            method = "sigdiff"
            path_type = "torch"
        else:
            # Operation not supported
            return None

        if kernel is None:
            return None

        # Run manual timing loop
        t_ms = self.manual_timing_loop(kernel)

        # Format and return result
        return self.output_result(
            t_ms=t_ms,
            library="chen-signatures",
            method=method,
            path_type=path_type,
            language="python"
        )


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: run_chen.py '<json_config>'", file=sys.stderr)
        sys.exit(1)

    # Parse configuration from command line
    config = json.loads(sys.argv[1])

    # Create and run adapter
    adapter = ChenSignaturesAdapter(config)
    adapter.run()

File: adapters/julia/run_chen.jl
#!/usr/bin/env julia
# ChenSignatures.jl adapter for signature benchmarks

using ChenSignatures
using StaticArrays
using JSON

# -------- Path Generators --------

"""Generate linear path: [t, 2t, 2t, ...]"""
function make_path_linear(d::Int, N::Int)
    ts = range(0.0, stop=1.0, length=N)
    [SVector{d,Float64}(ntuple(i -> (i == 1 ? t : 2t), d)) for t in ts]
end

"""Generate sinusoidal path: [sin(2π·1·t), sin(2π·2·t), ...]"""
function make_path_sin(d::Int, N::Int)
    ts = range(0.0, stop=1.0, length=N)
    ω = 2π
    [SVector{d,Float64}(ntuple(i -> sin(ω * i * t), d)) for t in ts]
end

"""Generate path of specified kind"""
function make_path(d::Int, N::Int, kind::String)
    kind_lower = lowercase(kind)
    if kind_lower == "linear"
        return make_path_linear(d, N)
    elseif kind_lower == "sin"
        return make_path_sin(d, N)
    else
        error("Unknown path_kind: $kind")
    end
end

# -------- Manual Timing Loop --------

"""
Execute manual timing loop with warmup and GC disabled.

This is the Julia equivalent of the Python BenchmarkAdapter.manual_timing_loop.

Returns a tuple of (avg_time_ms, avg_alloc_bytes):
    - avg_time_ms: Average time per iteration in milliseconds
    - avg_alloc_bytes: Average bytes allocated per iteration
"""
function manual_timing_loop(func::Function, repeats::Int; warmup_iterations::Int=3)
    # Warmup phase (untimed)
    for _ in 1:warmup_iterations
        func()
    end

    # Timed phase with GC disabled and allocation tracking
    GC.enable(false)
    local t0, t1, total_alloc_bytes
    try
        t0 = time_ns()

        # Track allocations for all iterations
        total_alloc_bytes = @allocated begin
            for _ in 1:repeats
                func()
            end
        end

        t1 = time_ns()
    finally
        GC.enable(true)
    end

    # Calculate average time in milliseconds
    total_time_ns = t1 - t0
    avg_time_ms = (total_time_ns / repeats) / 1e6

    # Calculate average bytes allocated per iteration
    avg_alloc_bytes = div(total_alloc_bytes, repeats)

    return avg_time_ms, avg_alloc_bytes
end

# -------- Benchmark Operations --------

"""
Prepare signature computation kernel.

Returns a closure that performs only the kernel (no setup).
"""
function run_signature(path, m::Int)
    # Setup phase (untimed): path is already prepared
    tensor_type = ChenSignatures.Tensor{Float64}

    # Return kernel closure
    return () -> signature_path(tensor_type, path, m)
end

"""
Prepare logsignature computation kernel.

Returns a closure that performs only the kernel (no setup).
"""
function run_logsignature(path, m::Int)
    # Setup phase (untimed): path is already prepared
    tensor_type = ChenSignatures.Tensor{Float64}

    # Return kernel closure
    return () -> ChenSignatures.log(signature_path(tensor_type, path, m))
end

# -------- Main Benchmark Runner --------

"""Format benchmark result for JSON output"""
function output_result(N, d, m, path_kind, operation, t_ms, alloc_bytes, library, method, path_type)
    return Dict(
        "N" => N,
        "d" => d,
        "m" => m,
        "path_kind" => path_kind,
        "operation" => operation,
        "language" => "julia",
        "library" => library,
        "method" => method,
        "path_type" => path_type,
        "t_ms" => t_ms,
        "alloc_bytes" => alloc_bytes
    )
end

"""Run the benchmark and output result as JSON"""
function run_benchmark(config::Dict)
    # Extract configuration
    N = config["N"]
    d = config["d"]
    m = config["m"]
    path_kind = config["path_kind"]
    operation = config["operation"]
    repeats = config["repeats"]

    # Generate path
    path = make_path(d, N, path_kind)

    # Select operation and prepare kernel
    if operation == "signature"
        kernel = run_signature(path, m)
        method = "signature_path"
    elseif operation == "logsignature"
        kernel = run_logsignature(path, m)
        method = "log"
    else
        # Operation not supported (sigdiff not available yet)
        error("Unsupported operation: $operation")
    end

    # Run manual timing loop
    t_ms, alloc_bytes = manual_timing_loop(kernel, repeats)

    # Format result
    result = output_result(
        N, d, m, path_kind, operation, t_ms, alloc_bytes,
        "ChenSignatures.jl",
        method,
        "Vector{SVector}"
    )

    # Output as JSON
    println(JSON.json(result))
end

# -------- Entry Point --------

if isempty(ARGS)
    println(stderr, "Usage: run_chen.jl '<json_config>'")
    exit(1)
end

# Parse configuration from command line
config_json = JSON.parse(ARGS[1])
# Convert JSON.Object to Dict
config = Dict{String, Any}(config_json)

# Run benchmark
try
    run_benchmark(config)
catch e
    error_result = Dict(
        "error" => string(e),
        "N" => get(config, "N", nothing),
        "d" => get(config, "d", nothing),
        "m" => get(config, "m", nothing),
        "operation" => get(config, "operation", nothing)
    )
    println(stderr, JSON.json(error_result))
    exit(1)
end
